### PR TITLE
State DB and snapshot compression

### PR DESCRIPTION
**Pull request by keorn**
Replaces common long RLPs with 2 byte invalid ones ( #177). When RLP is inside data RLP, then it is unpacked, compressed and marked. Two main additions are `Compressible` trait on `UntrustedRlp` and `HashDB` wrapper `CompressedDB`.
Needs more tests.

### CODE REVIEW

**Comment by keorn** [comment]
Looking at raw bytes state goes from 162MB to 130MB with `--pruning=fast` at block 1,440,000.

File: util/src/compresseddb.rs [hunk 1]
```diff
@@ -110,18 +111,19 @@ impl<'db> HashDB for CompressedDBMut<'db> {
 	fn insert(&mut self, value: &[u8]) -> H256 {
 		let key = value.sha3();
 		self.backing.emplace(key,
-												 UntrustedRlp::new(value).compress()
+												 UntrustedRlp::new(value)
+												 .compress()
```
Comment (rphmeier) [inline util/src/compresseddb.rs]: indentation?

File: util/src/compresseddb.rs [hunk 2]
```diff
@@ -110,18 +111,19 @@ impl<'db> HashDB for CompressedDBMut<'db> {
...
 												 .map(|r| r.to_vec())
 												 .unwrap_or(value.to_vec()));
 		key
 	}
 
 	fn emplace(&mut self, key: H256, value: Bytes) {
 		self.backing.emplace(key,
-												 UntrustedRlp::new(&value).compress()
+												 UntrustedRlp::new(&value)
+												 .compress()
```
Comment (rphmeier) [inline util/src/compresseddb.rs]: indentation?

**Comment by keorn** [comment]
Importing blocks up to 500,000 takes 65min instead of 60min according to a single trial.

File: util/src/rlp/untrusted_rlp.rs [hunk 3]
```diff
@@ -193,7 +197,7 @@ impl<'a, 'view> View<'a, 'view> for UntrustedRlp<'a> where 'a: 'view {
 	fn size(&self) -> usize {
 		match self.is_data() {
 			// we can safely unwrap (?) cause its data
-			true => BasicDecoder::payload_info(self.bytes).unwrap().value_len,
+			true => BasicDecoder::payload_info(self.bytes).map(|b| b.value_len).unwrap_or(0),
```
Comment (rphmeier) [inline util/src/rlp/untrusted_rlp.rs]: outdated comment here?

File: util/src/rlp/untrusted_rlp.rs [hunk 4]
```diff
@@ -55,6 +55,18 @@ pub struct PayloadInfo {
 	pub value_len: usize,
 }
 
+fn calculate_payload_info(header_bytes: &[u8], len_of_len: &usize) -> Result<PayloadInfo, DecoderError> {
```
Comment (rphmeier) [inline util/src/rlp/untrusted_rlp.rs]: reference to `usize` is the same size as `usize` itself -- better to pass-by-value

File: ethcore/src/account.rs [hunk 5]
```diff
@@ -120,7 +120,7 @@ impl Account {
 	}
 
 	/// Get (and cache) the contents of the trie's storage at `key`.
-	pub fn storage_at(&self, db: &AccountDB, key: &H256) -> H256 {
+	pub fn storage_at(&self, db: &HashDB, key: &H256) -> H256 {
```
Comment (rphmeier) [inline ethcore/src/account.rs]: reason for this change? `AccountDB` provides a few more guarantees than just `HashDB` and it doesn't look like you've changed the call sites

File: ethcore/src/state.rs [hunk 6]
```diff
@@ -177,7 +179,7 @@ impl State {
 
 	/// Mutate storage of account `address` so that it is `value` for `key`.
 	pub fn storage_at(&self, address: &Address, key: &H256) -> H256 {
-		self.get(address, false).as_ref().map_or(H256::new(), |a|a.storage_at(&AccountDB::new(self.db.as_hashdb(), address), key))
+		self.get(address, false).as_ref().map_or(H256::new(), |a| a.storage_at(&AccountDB::new(self.db.as_hashdb(), address), key))
```
Comment (rphmeier) [inline ethcore/src/state.rs]: not calling `CompressedDB:::new` here?

File: ethcore/src/account.rs (hunk 5 above)
Comment (keorn) [inline ethcore/src/account.rs]: The call sites do not use the additional `AccountDB` features, so might as well use simpler `HashDB`.

File: util/src/rlp/untrusted_rlp.rs (hunk 3 above)
Comment (keorn) [inline util/src/rlp/untrusted_rlp.rs]: This just removes a panic, does not really handle any errors properly (that needs some bigger refactor). Will change the comment to reflect that.

File: ethcore/src/state.rs (hunk 6 above)
Comment (keorn) [inline ethcore/src/state.rs]: Citing @arkpar "Smart contracts don't use RLP. There's nearly 0% chance you'll get an RLP in the contract storage"

**Comment by gavofyork** [comment]
i'm measuring a greater slowdown: around 40-50%. i suspect making database compression optional is going to be needed.

**Comment by keorn** [comment]
makes sense, might be faster for blocks db.

**Comment by gavofyork** [comment]
140s vs 247s:

`./parity.master import ~/first200k.rlp -d /tmp/master2  721.43s user 14.22s system 523% cpu 2:20.45 total`

`./parity.compressed import ~/first200k.rlp -d /tmp/compressed2  802.97s user 17.65s system 331% cpu 4:07.18 total`

**Comment by gavofyork** [comment]
looks like it's about 12% more work and that it kills parallelism by ~60%.

**Comment by keorn** [comment]
Thanks, will have a look.

File: ethcore/src/account.rs (hunk 5 above)
Comment (rphmeier) [inline ethcore/src/account.rs]: passing anything other than an AccountDB could introduce bugs or database key conflicts. imo it is better to leverage a strong type system to preserve invariants like that

**Comment by gavofyork** [comment]
Used updated PR, now 140s vs 194s:

```
./parity.compressed import ~/first200k.rlp -d /tmp/nullcompressdb2  769.20s user 14.14s system 403% cpu 3:14.14 total
```

versus:

```
./parity.master import ~/first200k.rlp -d /tmp/master2 721.43s user 14.22s system 523% cpu 2:20.45 total
```

**Comment by gavofyork** [comment]
Also led to a consensus failure:

```
     Running `target/release/parity`
Migrating database from version 7 to 8
Migration finished
2016-07-15 16:41:25 Starting Parity/v1.3.0-unstable-ee3df43-20160715/x86_64-macos/rustc1.9.0
2016-07-15 16:41:25 Using state DB journalling strategy fast
2016-07-15 16:41:25 Configured for Frontier/Homestead using Ethash engine
2016-07-15 16:41:29 Received old authentication request.
2016-07-15 16:41:29 Unauthorized connection to Signer API blocked.
2016-07-15 16:41:34 Stage 4 block verification failed for #1889987 (b2fa…c4a5)
Error: Block(InvalidGasUsed(Mismatch { expected: 436479, found: 503579 }))
#1889986 58d5…a9aa      0 blk/s    0 tx/s   0 Mgas/s    0/ 4/25 peers   #1890068     0+    0 Qed     49 MiB db   10 KiB chain    2 KiB queue   20 KiB sync
```

**Comment by gavofyork** [comment]
possibly some issue with trying to decompress uncompressed data from the database?

**Comment by keorn** [comment]
I will remove state compression for now as it will require something like a write queue optimization. Instead will do a blocks+snapshot pr.

**Pull request closed without merging**